Support `features.storyStoreV7`

joshwooding opened the pull request:
Closes #125

IanVS commented:
Hi @joshwooding, thanks for your work on this.  Is there more you're planning to do, or is it ready to review?  If so, is there anything we should be aware of as we do?

joshwooding commented:
@IanVS I think there are some few kinks to iron out. Currently the current issues are:

- Storybook uses the prebuilt manager which breaks everything. You need to delete node_modules > @storybook/manager-webpack4 > prebuild to get everything working or use --no-manager-cache (This might be an issue with the monorepo)
- You need to refresh the page when initial loading because a 408 timeout is returned when getting @react-refresh

tmeasday commented on packages/storybook-builder-vite/codegen-importfn-script.js:
```diff
@@ -0,0 +1,48 @@
+          return;
+        }
+        
+        const pathRemainder = path.substring(${directory.length + 1});
+        const module = (await import('${storyImports}')).default;
+        return module[path]();
```
@joshwooding can you tell me a bit about this snippet? Does `storyImports` contain a straight hash of filename => import functions?

tmeasday commented on packages/storybook-builder-vite/codegen-importfn-script.js:
```diff
@@ -0,0 +1,48 @@
+          return;
+        }
+        
+        const pathRemainder = path.substring(${directory.length + 1});
+        const module = (await import('${storyImports}')).default;
+        return module[path]();
```
@joshwooding I think you could cut out the middle man here and just put your hash from `codegen-virtual-imports-script.js` in here directly:

```js
const importers = {
  'x': async() => import('x'), // one per CSF file, as you do below
}

const toImportFn = async (path) => importers[path](); 
```

We might eventually do something similar in webpack too, in which case we should definitely try and share code for this generation,.

The complex code we have for the webpack (`importFnPart` etc) is to trigger webpack's glob matching behaviour which you don't need AFIACT (?)

joshwooding replied:
Great idea. This all seems to work to me. I was trying to stick closely to how webpack does it to avoid confusion but it's much cleaner this way.

joshwooding commented:
@IanVS I'm stumped on the 408 timeout issue. It seems like the modules referenced in the iframe aren't loaded on the first render so we get a 408 error in the console and the stories don't load. Refreshing the page fixes this. Any ideas?

eirslett commented:
> @eirslett Not sure if you have any ideas?

We usually add any third-party dependencies from package.json to the `optimizeDeps` config, so they are optimized immediately on startup.

joshwooding commented:
@eirslett The 408 requests include third-party dependencies sadly :(

![image](https://user-images.githubusercontent.com/12938082/140066657-dd26dd18-608c-4a57-9050-f4eba5757890.png)

joshwooding commented:
The only issue I can see now is that the static storybook sets an empty path?

joshwooding commented:
> The only issue I can see now is that the static storybook sets an empty path?

This is an issue with `storybook-builder-vite` so as far as I can see this is ready to review @IanVS 🎉

IanVS commented:
Thanks for sticking with this one, @joshwooding!  I'll give it some 👀 as soon as I can.  

In the meantime, it would be great if anyone else would be willing to try it out and play around with it, especially those using svelte or vue or other frameworks than React, which is what I'll be testing out.

joshwooding commented:
Hmm, I can't get svelte to work in the repo. Two instances of ClientApi seem to exist for some reason

joshwooding commented:
.svelte story files no longer work due to https://github.com/storybookjs/storybook/blob/next/lib/core-server/src/utils/StoryIndexGenerator.ts#L68 @tmeasday Do you know if this is intentional?

tmeasday commented:
Hmm, yes I guess we will need to think about that @shilman

joshwooding commented on packages/example-react/package.json:
```diff
@@ -5,7 +5,7 @@
     "main": "index.js",
     "scripts": {
-        "storybook": "start-storybook",
+        "storybook": "start-storybook --no-manager-cache",
```
https://github.com/storybookjs/storybook/pull/16653 should fix the issue

joshwooding replied:
I've tested it locally and it still seems to need --no-manager-cache :/

tmeasday replied:
@joshwooding can you send a repro?

joshwooding replied:
@tmeasday I'm still seeing the same issue using the repo in https://github.com/storybookjs/storybook/issues/16647

joshwooding replied:
Can confirm this is now fixed 🎉

joshwooding commented:
> .svelte story files no longer work due to https://github.com/storybookjs/storybook/blob/next/lib/core-server/src/utils/StoryIndexGenerator.ts#L68 cc @tmeasday
> 
> I assume this is intentional
> 
> [storybookjs/storybook#16673](https://github.com/storybookjs/storybook/issues/16673) an issue has been raised to track this.

I think this is the only issue left :)

IanVS commented on packages/example-react/.storybook/main.js:
```diff
@@ -12,6 +12,9 @@ module.exports = {
         builder: 'storybook-builder-vite',
     },
+    features: {
+      storyStoreV7: true,
```
@joshwooding would you mind trying to enable addon-a11y?  I'm seeing an error in my own react app when I try to enable storyStoreV7 using that addon in recent versions of storybook.  (the addon has been working fine previously, and enabling the story store v7 did too, or at least it wasn't breaking anything in older beta versions, but does now.)

joshwooding replied:
Hmm, everything is working for me @IanVS What kind of behaviour are you seeing?

IanVS replied:
Ok cool.  Maybe some of the changes you've made here in this PR fix it.  I just thought it'd be worth checking out.  We should uncomment the line to make sure it keeps working.  

In my own app, I was seeing this at runtime:

```
index.js:30 Uncaught Error: Accessing non-existent addons channel, see https://storybook.js.org/basics/faq/#why-is-there-no-addons-channel
    at AddonStore2.getChannel (index.js:30)
    at a11yRunner.js? [sm]:19
```

But, that wasn't using this branch.

joshwooding replied:
I think uncommenting is a good idea :)

IanVS commented:
I pulled down the branch and am experimenting with it now.  I'm seeing a few things so far, still investigating:

```

╭──────────────────────────────────────────────────────╮
│                                                      │
│   Storybook 6.4.0-rc.2 for React started             │
│   6.56 s for manager and 5.38 s for preview          │
│                                                      │
│    Local:            http://localhost:62893/         │
│    On your network:  http://192.168.68.116:62893/    │
│                                                      │
╰──────────────────────────────────────────────────────╯
8:25:39 PM [vite] new dependencies found: @storybook/channel-websocket, updating...
Sourcemap for "/Users/ianvs/code/storybook/storybook-builder-vite/packages/example-react/node_modules/.vite/@storybook_preview-web.js" points to missing source files
Sourcemap for "/Users/ianvs/code/storybook/storybook-builder-vite/packages/example-react/node_modules/.vite/@storybook_client-api.js" points to missing source files
Sourcemap for "/Users/ianvs/code/storybook/storybook-builder-vite/packages/example-react/node_modules/.vite/global.js" points to missing source files
Sourcemap for "/Users/ianvs/code/storybook/storybook-builder-vite/packages/example-react/node_modules/.vite/@storybook_addons.js" points to missing source files
Sourcemap for "/Users/ianvs/code/storybook/storybook-builder-vite/packages/example-react/node_modules/.vite/@storybook_channel-postmessage.js" points to missing source files
8:25:39 PM [vite] Failed to load source map for /node_modules/.vite/chunk-JST3G223.js?v=732bfa54.
8:25:39 PM [vite] Failed to load source map for /node_modules/.vite/chunk-XEMAM7KF.js?v=732bfa54.
8:25:39 PM [vite] Failed to load source map for /node_modules/.vite/chunk-MRP3MKNT.js?v=732bfa54.
8:25:39 PM [vite] Failed to load source map for /node_modules/.vite/chunk-TYR6E6LN.js?v=732bfa54.
8:25:39 PM [vite] Failed to load source map for /node_modules/.vite/chunk-NRERFRFO.js?v=732bfa54.
8:25:39 PM [vite] Failed to load source map for /node_modules/.vite/chunk-OAV7HHZ3.js?v=732bfa54.
8:25:39 PM [vite] Failed to load source map for /node_modules/.vite/chunk-LRSF2QUR.js?v=732bfa54.
.
.
.
Sourcemap for "/Users/ianvs/code/storybook/storybook-builder-vite/packages/example-react/node_modules/.vite/unfetch.js" points to missing source files
8:25:41 PM [vite] ✨ dependencies updated, reloading page...
Sourcemap for "/Users/ianvs/code/storybook/storybook-builder-vite/packages/example-react/node_modules/.vite/@storybook_client-logger.js" points to missing source files
Sourcemap for "/Users/ianvs/code/storybook/storybook-builder-vite/packages/example-react/node_modules/.vite/telejson.js" points to missing source files
Sourcemap for "/Users/ianvs/code/storybook/storybook-builder-vite/packages/example-react/node_modules/.vite/chunk-4S6UEONU.js" points to missing source files
Sourcemap for "/Users/ianvs/code/storybook/storybook-builder-vite/packages/example-react/node_modules/.vite/util-deprecate.js" points to missing source files
Sourcemap for "/Users/ianvs/code/storybook/storybook-builder-vite/packages/example-react/node_modules/.vite/ts-dedent.js" points to missing source files
Sourcemap for "/Users/ianvs/code/storybook/storybook-builder-vite/packages/example-react/node_modules/.vite/chunk-AKDXD5X2.js" points to missing source files
```

So, it seems like we should add `@storybook/channel-websocket` to the `optimizeDeps` list, and I've seen others report the source map issues, so now that I have a reproduction I'll see if I can figure out what's going on.

IanVS commented:
It's so cool to see the new components get fetched only when you view their stories.  This should help folks with large storybooks a ton!

IanVS reviewed (commented):
If you're willing to add `@storybook/channel-websocket` to the list of optimized dependencies and uncomment out the accessibility plugin in the examples, I'll give this my :+1:.  Great work here, thanks again for sticking with it.

tmeasday commented:
> It's so cool to see the new components get fetched only when you view their stories

In Vite I would assume they are only getting compiled when requested too, which I assume should help larger storybooks a lot!

IanVS commented:
Compiling is really fast in vite, but one of the bottlenecks has been sending all the (sometimes hundreds of) files to the browser at startup.  So for both reasons, this is a big step in the right direction!

joshwooding commented on packages/storybook-builder-vite/codegen-iframe-script.js:
```diff
@@ -95,6 +94,10 @@ module.exports.generateIframeScriptCode =
+          case 'decorateStory':
+          case 'renderToDOM': {
+            return null; // This key is not handled directly in v6 mode.
+          }
```
There's still one about render which might be fixed by adding: https://github.com/storybookjs/storybook/blob/next/lib/builder-webpack4/src/preview/virtualModuleEntry.template.js#L35

Didn't want to add it in this PR though. Will investigate after.

joshwooding commented:
> It's so cool to see the new components get fetched only when you view their stories. This should help folks with large storybooks a ton!
> 
> What do you think about merging this and in our release notes mentioning that `storyStoreV7` is not supported for svelte yet. It sounds like storybook changes will be needed for it to work correctly, so there's probably no need to hold up this PR. This whole project is experimental, so I'm 👍 for releasing and getting bug reports if folks hit trouble.

Sounds great. I've added comments and updated optimizeDeps :)

IanVS commented:
Oh shoot, I've caused a merge conflict in the lockfile.  Would you mind rebasing one more time?

joshwooding commented:
@IanVS Done :)
